clash-lib-1.8.1 failed to build with ghc-9.8.4 Stackage Nightly

```
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:516:20: error: [GHC-83865]                                                                                       
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’                                                                       
             Expected: Clash.Unique.Unique                                                                                          
               Actual: GHC.Word.Word64                                                                                              
           • In the second argument of ‘(==)’, namely                                                                               
               ‘getKey typeNatAddTyFamNameKey’                                                                                      
             In the expression: nameUniq tc == getKey typeNatAddTyFamNameKey                                                        
             In a stmt of a pattern guard for                                                                                       
                            an equation for ‘reduceTypeFamily’:                                                                     
               nameUniq tc == getKey typeNatAddTyFamNameKey                                                                         
           |                                                                                                                        
       516 |   | nameUniq tc == getKey typeNatAddTyFamNameKey                                                                       
           |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                       
                                                                                                                                    
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:521:20: error: [GHC-83865]                                                                                       
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’                                                                       
             Expected: Clash.Unique.Unique                                                                                          
               Actual: GHC.Word.Word64                                                                                              
           • In the second argument of ‘(==)’, namely
               ‘getKey typeNatMulTyFamNameKey’
             In the expression: nameUniq tc == getKey typeNatMulTyFamNameKey
             In a stmt of a pattern guard for
                            an equation for ‘reduceTypeFamily’:
               nameUniq tc == getKey typeNatMulTyFamNameKey
           |
       521 |   | nameUniq tc == getKey typeNatMulTyFamNameKey
           |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:526:20: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the second argument of ‘(==)’, namely
               ‘getKey typeNatExpTyFamNameKey’
             In the expression: nameUniq tc == getKey typeNatExpTyFamNameKey
             In a stmt of a pattern guard for
                            an equation for ‘reduceTypeFamily’:
                            an equation for ‘reduceTypeFamily’:                                                01:50:14 [313/937727]
               nameUniq tc == getKey typeNatExpTyFamNameKey                                                                         
           |
       526 |   | nameUniq tc == getKey typeNatExpTyFamNameKey
           |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:531:20: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the second argument of ‘(==)’, namely
               ‘getKey typeNatSubTyFamNameKey’
             In the expression: nameUniq tc == getKey typeNatSubTyFamNameKey
             In a stmt of a pattern guard for
                            an equation for ‘reduceTypeFamily’:
               nameUniq tc == getKey typeNatSubTyFamNameKey
           |
       531 |   | nameUniq tc == getKey typeNatSubTyFamNameKey
           |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:552:20: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the second argument of ‘(==)’, namely
               ‘getKey typeNatCmpTyFamNameKey’
             In the expression: nameUniq tc == getKey typeNatCmpTyFamNameKey
             In a stmt of a pattern guard for
                            an equation for ‘reduceTypeFamily’:
               nameUniq tc == getKey typeNatCmpTyFamNameKey
           |
       552 |   | nameUniq tc == getKey typeNatCmpTyFamNameKey -- "GHC.TypeNats.CmpNat"
           |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:557:45: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordLTDataConKey)’
             In the expression:
               Name User "GHC.Types.LT" (getKey ordLTDataConKey) wiredInSrcSpan
             In a case alternative:
                 LT
                   -> Name User "GHC.Types.LT" (getKey ordLTDataConKey) wiredInSrcSpan
           |
       557 |             LT -> Name User "GHC.Types.LT" (getKey ordLTDataConKey) wiredInSrcSpan
           |                                             ^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:558:45: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordEQDataConKey)’
             In the expression:
               Name User "GHC.Types.EQ" (getKey ordEQDataConKey) wiredInSrcSpan
             In a case alternative:
                 EQ
                   -> Name User "GHC.Types.EQ" (getKey ordEQDataConKey) wiredInSrcSpan
           |
       558 |             EQ -> Name User "GHC.Types.EQ" (getKey ordEQDataConKey) wiredInSrcSpan
           |                                             ^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:559:45: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordGTDataConKey)’
             In the expression:
               Name User "GHC.Types.GT" (getKey ordGTDataConKey) wiredInSrcSpan
             In a case alternative:
                 GT
                   -> Name User "GHC.Types.GT" (getKey ordGTDataConKey) wiredInSrcSpan
           |
       559 |             GT -> Name User "GHC.Types.GT" (getKey ordGTDataConKey) wiredInSrcSpan
           |                                             ^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:562:20: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the second argument of ‘(==)’, namely
               ‘getKey typeSymbolCmpTyFamNameKey’
             In the expression: nameUniq tc == getKey typeSymbolCmpTyFamNameKey
             In a stmt of a pattern guard for
                            an equation for ‘reduceTypeFamily’:
               nameUniq tc == getKey typeSymbolCmpTyFamNameKey
           |
       562 |   | nameUniq tc == getKey typeSymbolCmpTyFamNameKey -- "GHC.TypeNats.CmpSymbol"
           |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:567:45: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordLTDataConKey)’
             In the expression:
               Name User "GHC.Types.LT" (getKey ordLTDataConKey) wiredInSrcSpan
             In a case alternative:
                 LT
                   -> Name User "GHC.Types.LT" (getKey ordLTDataConKey) wiredInSrcSpan
           |
       567 |             LT -> Name User "GHC.Types.LT" (getKey ordLTDataConKey) wiredInSrcSpan
           |                                             ^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:568:45: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordEQDataConKey)’
             In the expression:
               Name User "GHC.Types.EQ" (getKey ordEQDataConKey) wiredInSrcSpan
             In a case alternative:
                 EQ
                   -> Name User "GHC.Types.EQ" (getKey ordEQDataConKey) wiredInSrcSpan
           |
       568 |             EQ -> Name User "GHC.Types.EQ" (getKey ordEQDataConKey) wiredInSrcSpan
           |                                             ^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:569:45: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordGTDataConKey)’
             In the expression:
               Name User "GHC.Types.GT" (getKey ordGTDataConKey) wiredInSrcSpan
             In a case alternative:
                 GT
                   -> Name User "GHC.Types.GT" (getKey ordGTDataConKey) wiredInSrcSpan
           |
       569 |             GT -> Name User "GHC.Types.GT" (getKey ordGTDataConKey) wiredInSrcSpan
           |                                             ^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:573:20: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the second argument of ‘(==)’, namely
               ‘getKey typeCharCmpTyFamNameKey’
             In the expression: nameUniq tc == getKey typeCharCmpTyFamNameKey
             In a stmt of a pattern guard for
                            an equation for ‘reduceTypeFamily’:
               nameUniq tc == getKey typeCharCmpTyFamNameKey
           |
       573 |   | nameUniq tc == getKey typeCharCmpTyFamNameKey -- "GHC.TypeNats.CmpSymbol"
           |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:578:42: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordLTDataConKey)’
             In the expression:
               Name User (showt 'LT) (getKey ordLTDataConKey) wiredInSrcSpan
             In a case alternative:
                 LT -> Name User (showt 'LT) (getKey ordLTDataConKey) wiredInSrcSpan
           |
       578 |             LT -> Name User (showt 'LT) (getKey ordLTDataConKey) wiredInSrcSpan
           |                                          ^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:579:42: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordEQDataConKey)’
             In the expression:
               Name User (showt 'EQ) (getKey ordEQDataConKey) wiredInSrcSpan
             In a case alternative:
                 EQ -> Name User (showt 'EQ) (getKey ordEQDataConKey) wiredInSrcSpan
           |
       579 |             EQ -> Name User (showt 'EQ) (getKey ordEQDataConKey) wiredInSrcSpan
           |                                          ^^^^^^^^^^^^^^^^^^^^^^
        
       /home/curators/work/unpack-dir/unpacked/clash-lib-1.8.1-03ec2cb411fe7ac811914edffea21995ffd0d5c5b6898264fc548c766d146ecb/src/
Clash/Core/Type.hs:580:42: error: [GHC-83865]
           • Couldn't match type ‘GHC.Word.Word64’ with ‘Int’
             Expected: Clash.Unique.Unique
               Actual: GHC.Word.Word64
           • In the third argument of ‘Name’, namely
               ‘(getKey ordGTDataConKey)’
             In the expression:
               Name User (showt 'GT) (getKey ordGTDataConKey) wiredInSrcSpan
:
:
```

Not sure what changed

--- a/.ci/build.sh
+++ b/.ci/build.sh
@@ -31,7 +31,6 @@ TESTS="
 clash-cores:doctests
 clash-cores:unittests
 clash-cosim:test
-clash-ffi:ffi-interface-tests
 clash-lib:doctests
 clash-lib:unittests
 clash-prelude:doctests
@@ -42,3 +41,12 @@ mkdir bin
 for TEST in $TESTS; do
   ln -s "$(realpath --relative-to=bin "$(cabal list-bin $TEST)")" bin/$TEST
 done
+
+# TODO: remove this and put it back into tests when
+# https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12264#note_602406
+# is fixed
+set +u
+if [[ "$SKIP_CLASH_FFI_EXAMPLE" != "yes" ]]; then
+  ln -s "$(realpath --relative-to=bin "$(cabal list-bin clash-ffi:ffi-interface-tests)")" bin/clash-ffi:ffi-interface-tests
+fi
+set -u

--- a/.ci/docker/Dockerfile
+++ b/.ci/docker/Dockerfile
@@ -111,7 +111,7 @@ RUN git clone https://github.com/verilator/verilator verilator \
 
 FROM builder AS build-ghc
 
-ARG ghcup_version="0.1.18.0"
+ARG ghcup_version="0.1.30.0"
 
 # Must be explicitly set
 ARG ghc_version

--- a/.ci/docker/build-and-publish-docker-image.sh
+++ b/.ci/docker/build-and-publish-docker-image.sh
@@ -15,8 +15,8 @@ elif [[ "$1" != "" ]]; then
 fi
 
 UBUNTU_VERSION=jammy-20240808
-GHC_VERSIONS=("9.10.1" "9.8.2"  "9.6.6"  "9.4.8"  "9.2.8"   "9.0.2"   "8.10.7")
-CABAL_VERSION="3.12.1.0"
+GHC_VERSIONS=("9.10.1" "9.8.4"  "9.6.6"  "9.4.8"  "9.2.8"   "9.0.2"   "8.10.7")
+CABAL_VERSION="3.14.1.0"
 
 # We want to use docker buildkit so that our layers are built in parallel. This
 # is ignored completely on versions of docker which don't support buildkit.

--- a/.ci/gitlab/benchmark.yml
+++ b/.ci/gitlab/benchmark.yml
@@ -1,5 +1,5 @@
 .benchmark:
-  image: ghcr.io/clash-lang/clash-ci:$GHC_VERSION-20240721
+  image: ghcr.io/clash-lang/clash-ci:$GHC_VERSION-20250101
   stage: test
   timeout: 2 hours
   variables:

--- a/.ci/gitlab/common.yml
+++ b/.ci/gitlab/common.yml
@@ -12,7 +12,7 @@ default:
   timeout: 10 minutes
   stage: build
   variables:
-    CLASH_DOCKER_TAG: 20240905
+    CLASH_DOCKER_TAG: 20250101
     CACHE_BUST_TOKEN: 3
     # Note that we copy+paste the image name into CACHE_FALLBACK_KEY. If we don't,
     # $GHC_VERSION gets inserted at verbatim, instead of resolving to some ghc version.

--- a/.ci/gitlab/test.yml
+++ b/.ci/gitlab/test.yml
@@ -128,6 +128,10 @@ lib:unittests:
 
 ffi:interface-tests:
   extends: .test-nocache
+  # TODO: remove this when https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12264#note_602406
+  # is fixed
+  rules:
+    - if: $SKIP_CLASH_FFI_EXAMPLE != "yes"
   script:
     - bin/clash-ffi:ffi-interface-tests --smallcheck-max-count 2000
 
@@ -172,6 +176,10 @@ ffi:example:
   extends:
     - .common-local
     - .nightly-rules
+  # TODO: remove this when https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12264#note_602406
+  # is fixed
+  rules:
+    - if: $SKIP_CLASH_FFI_EXAMPLE != "yes"
   stage: test
   needs: []
   before_script:

--- a/.ci/stack-9.8.yaml
+++ b/.ci/stack-9.8.yaml
@@ -1,4 +1,4 @@
-resolver: nightly-2024-07-23
+resolver: lts-23.2
 
 ghc-options:
   "$locals": -Wall -Wcompat
@@ -16,9 +16,5 @@ extra-deps:
 - fakedata-1.0.3@sha256:75630f6b9665965b4b551239ffcf89e4f879f93d235b18d4b249b0ed50513d0d,24593
 - hedgehog-fakedata-0.0.1.5@sha256:d8059e4ef9b7b4112bef9791300118f3a2d776bb191e50b41635a411af609428,1424
 - string-random-0.1.4.3@sha256:4b3092568889e00b419093a1938f14a67743844bc86476ee5cb6f8827dc40545,2172
-# TODO: Remove this workaround. See:
-#
-#   https://github.com/clash-lang/clash-compiler/pull/2665#issuecomment-1939044550
-- Cabal-3.10.3.0
 
 allow-newer: true

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -154,7 +154,7 @@ jobs:
 
     # Run steps inside the clash CI docker image
     container:
-      image: ghcr.io/clash-lang/clash-ci:${{ matrix.ghc }}-20240721
+      image: ghcr.io/clash-lang/clash-ci:${{ matrix.ghc }}-20250101
 
       env:
         THREADS: 2

--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -38,6 +38,9 @@ tests:
         RUN_TESTS: "always"
 
       - GHC_VERSION: 9.8.4
+        # TODO: remove this when https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12264#note_602406
+        # is fixed
+        SKIP_CLASH_FFI_EXAMPLE: "yes"
         RUN_TESTS: "always"
 
       - GHC_VERSION: 9.6.6

--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -37,7 +37,7 @@ tests:
       - GHC_VERSION: 9.10.1
         RUN_TESTS: "always"
 
-      - GHC_VERSION: 9.8.2
+      - GHC_VERSION: 9.8.4
         RUN_TESTS: "always"
 
       - GHC_VERSION: 9.6.6

--- a/changelog/2024-12-31T16_09_30+01_00_ghc_984_support
+++ b/changelog/2024-12-31T16_09_30+01_00_ghc_984_support
@@ -1,0 +1,1 @@
+ADDED: Support for GHC 9.8.4 [#2852](https://github.com/clash-lang/clash-compiler/issues/2852)

--- a/clash-ffi/clash-ffi.cabal
+++ b/clash-ffi/clash-ffi.cabal
@@ -76,6 +76,12 @@ library
     Clash.FFI.VPI.Port
     Clash.FFI.VPI.Port.Direction
     Clash.FFI.VPI.Reg
+  -- TODO: remove this once https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12264#note_602406
+  -- is fixed
+  if impl(ghc < 9.8.3) || impl(ghc == 9.10.1)
+    buildable: true
+  else
+    buildable: false
 
 executable ffi-interface-tests
   import: common-options
@@ -107,3 +113,9 @@ executable ffi-interface-tests
     , tasty
     , tasty-hunit
     , tasty-smallcheck
+  -- TODO: remove this once https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12264#note_602406
+  -- is fixed
+  if impl(ghc < 9.8.3) || impl(ghc == 9.10.1)
+    buildable: true
+  else
+    buildable: false

--- a/clash-lib-hedgehog/clash-lib-hedgehog.cabal
+++ b/clash-lib-hedgehog/clash-lib-hedgehog.cabal
@@ -61,5 +61,6 @@ library
     primitive                 >= 0.5.0.1 && < 1.0,
     text                      >= 1.2.2   && < 2.2,
     transformers              >= 0.5.2.0 && < 0.7,
+    ghc                       >= 8.10.0  && < 9.11,
 
     clash-lib                 == 1.9.0,

--- a/clash-lib-hedgehog/src/Clash/Hedgehog/Unique.hs
+++ b/clash-lib-hedgehog/src/Clash/Hedgehog/Unique.hs
@@ -35,7 +35,7 @@ import Clash.Hedgehog.Internal.Bias
 
 genUnique :: forall m. MonadGen m => m Unique
 genUnique =
-#if __GLASGOW_HASKELL__ >= 910
+#if MIN_VERSION_ghc(9,8,4)
   Gen.word64
 #else
   Gen.int

--- a/clash-lib/src/Clash/Core/FreeVars.hs
+++ b/clash-lib/src/Clash/Core/FreeVars.hs
@@ -33,7 +33,7 @@ import qualified Control.Lens           as Lens
 import Control.Lens.Fold                (Fold)
 import Control.Lens.Getter              (Contravariant)
 import Data.Coerce
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
 import qualified GHC.Data.Word64Set     as IntSet
 #else
 import qualified Data.IntSet            as IntSet
@@ -86,7 +86,7 @@ typeFreeVars'
   :: (Contravariant f, Applicative f)
   => (forall b . Var b -> Bool)
   -- ^ Predicate telling whether a variable is interesting
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
   -> IntSet.Word64Set
 #else
   -> IntSet.IntSet

--- a/clash-lib/src/Clash/Core/VarEnv.hs
+++ b/clash-lib/src/Clash/Core/VarEnv.hs
@@ -103,7 +103,7 @@ import           Data.Coerce               (coerce)
 import qualified Data.List                 as List
 import qualified Data.List.Extra           as List
 import           Data.Maybe                (fromMaybe)
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
 import           Data.Word                 (Word64)
 #endif
 
@@ -389,7 +389,7 @@ eltsVarSet = UniqMap.elems
 -- * InScopeSet
 
 type Seed
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
   = Word64
 #else
   = Int

--- a/clash-lib/src/Clash/Data/UniqMap.hs
+++ b/clash-lib/src/Clash/Data/UniqMap.hs
@@ -45,7 +45,7 @@ import           Control.DeepSeq (NFData)
 import           Data.Binary (Binary (..))
 import           Data.Bifunctor (first)
 import           Data.Function (on)
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
 import           GHC.Data.Word64Map.Strict (Word64Map)
 import qualified GHC.Data.Word64Map.Strict as IntMap
 #else
@@ -71,7 +71,7 @@ import           Clash.Unique (Unique, Uniquable(getUnique))
 -- uniqueable and provide their own key, however a unique can be associated
 -- with any value.
 newtype UniqMap a
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
   = UniqMap { uniqMapToIntMap :: Word64Map a }
 #else
   = UniqMap { uniqMapToIntMap :: IntMap a }
@@ -85,7 +85,7 @@ newtype UniqMap a
     , Semigroup
     , Show
     )
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
 instance Binary a => Binary (UniqMap a) where
   put (UniqMap m) = put (IntMap.size m) <> mapM_ put (IntMap.toAscList m)
   get             = fmap (UniqMap . IntMap.fromDistinctAscList) get

--- a/clash-lib/src/Clash/Netlist/Util.hs
+++ b/clash-lib/src/Clash/Netlist/Util.hs
@@ -43,11 +43,6 @@ import           Data.Functor            (($>))
 import           Data.Hashable           (Hashable)
 import           Data.HashMap.Strict     (HashMap)
 import qualified Data.HashMap.Strict     as HashMap
-#if MIN_VERSION_ghc(9,10,0)
-import qualified GHC.Data.Word64Set      as IntSet
-#else
-import qualified Data.IntSet             as IntSet
-#endif
 import           Data.Primitive.ByteArray (ByteArray (..))
 import           Control.Applicative     (Alternative((<|>)))
 import           Data.List               (unzip4, partition)
@@ -573,10 +568,10 @@ hasUnconstrainedExistential tcm dc =
         let -- Free FVs in the LHS and RHS of the constraint that are not the
             -- in the set of universal type variables of the constructor.
             ty1FEVs = Lens.toListOf (typeFreeVars' ((`notElem` uTVs) . coerce)
-                                                   IntSet.empty)
+                                                   mempty)
                                     ty1
             ty2FEVs = Lens.toListOf (typeFreeVars' ((`notElem` uTVs) . coerce)
-                                                   IntSet.empty)
+                                                   mempty)
                                     ty2
 
             -- Determine whether `eTV` can be generated from one side of a

--- a/clash-lib/src/Clash/Normalize/Transformations/DEC.hs
+++ b/clash-lib/src/Clash/Normalize/Transformations/DEC.hs
@@ -51,11 +51,6 @@ import qualified Data.Foldable as Foldable
 import qualified Data.Graph as Graph
 import Data.IntMap.Strict (IntMap)
 import qualified Data.IntMap.Strict as IntMap
-#if MIN_VERSION_ghc(9,10,0)
-import qualified GHC.Data.Word64Set as IntSet
-#else
-import qualified Data.IntSet as IntSet
-#endif
 import qualified Data.List as List
 import qualified Data.List.Extra as List
 import qualified Data.Map.Strict as Map
@@ -599,7 +594,7 @@ areShared _   _       []       = True
 areShared tcm inScope xs@(x:_) = noFV1 && (isProof x || allEqual xs)
  where
   noFV1 = case x of
-    Right ty -> getAll (Lens.foldMapOf (typeFreeVars' isLocallyBound IntSet.empty)
+    Right ty -> getAll (Lens.foldMapOf (typeFreeVars' isLocallyBound mempty)
                                        (const (All False)) ty)
     Left tm  -> getAll (Lens.foldMapOf (termFreeVars' isLocallyBound)
                                        (const (All False)) tm)

--- a/clash-lib/src/Clash/Unique.hs
+++ b/clash-lib/src/Clash/Unique.hs
@@ -18,7 +18,7 @@ module Clash.Unique
   ) where
 
 import Data.Word (Word64)
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
 import GHC.Word (Word64(W64#))
 import GHC.Exts (Word64#)
 #else
@@ -31,7 +31,7 @@ import qualified GHC.Types.Unique as GHC
 import qualified Unique as GHC
 #endif
 
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
 type Unique = Word64
 type Unique# = Word64#
 
@@ -63,16 +63,11 @@ instance Uniquable Unique where
   getUnique = id
   setUnique = flip const
 
-#if !MIN_VERSION_ghc(9,10,0)
+#if !MIN_VERSION_ghc(9,8,4)
 instance Uniquable Word64 where
   getUnique = fromIntegral
   setUnique _ = fromIntegral
 #endif
 
-#if MIN_VERSION_ghc(9,10,0)
 fromGhcUnique :: GHC.Unique -> Unique
-fromGhcUnique = fromIntegral . GHC.getKey
-#else
-fromGhcUnique :: GHC.Unique -> Unique
-fromGhcUnique = id . GHC.getKey
-#endif
+fromGhcUnique = GHC.getKey

--- a/clash-lib/src/Clash/Util/Graph.hs
+++ b/clash-lib/src/Clash/Util/Graph.hs
@@ -16,7 +16,7 @@ module Clash.Util.Graph
 
 import           Data.Tuple            (swap)
 import           Data.Foldable         (foldlM)
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
 import qualified GHC.Data.Word64Map.Strict as IntMap
 import qualified GHC.Data.Word64Set        as IntSet
 #else
@@ -33,7 +33,7 @@ import           Clash.Driver.Types (BindingMap, Binding (bindingTerm))
 import           Clash.Normalize.Util (callGraph)
 import           Clash.Unique (Unique)
 
-#if MIN_VERSION_ghc(9,10,0)
+#if MIN_VERSION_ghc(9,8,4)
 type IntMap = IntMap.Word64Map
 type IntSet = IntSet.Word64Set
 #endif

